mne/io/base.py

```diff
@@ -836,6 +836,50 @@ def __setitem__(self, item, value):
         # set the data
         self._data[sel, start:stop] = value
 
+    def _get_data(self, sel, start, stop):
+        """Get data omitting bad segments."""
+        if self.annotations is None:
+            return self[sel, start:stop]
+        sfreq = self.info['sfreq']
+        bads = [idx for idx, desc in enumerate(self.annotations.description)
+                if desc.upper().startswith('BAD')]
+        onsets = self.annotations.onset[bads]
+        onsets = _sync_onset(self, onsets)
+        ends = onsets + self.annotations.duration[bads]
+        omit = np.concatenate([np.where(onsets > stop / sfreq)[0],
+                               np.where(ends < start / sfreq)[0]])
+        onsets, ends = np.delete(onsets, omit), np.delete(ends, omit)
+        if len(onsets) == 0:
+            return self[sel, start:stop]
+        stop = min(stop, self.n_times)
+        order = np.argsort(onsets)
+        onsets = self.time_as_index(onsets[order])
+        ends = self.time_as_index(ends[order])
+
+        np.clip(onsets, start, stop, onsets)
+        np.clip(ends, start, stop, ends)
+        used = np.ones(stop - start, bool)
+        for onset, end in zip(onsets, ends):
+            if onset >= end:
+                continue
+            used[onset - start: end - start] = False
+        used = np.concatenate([[False], used, [False]])
+        starts = np.where(~used[:-1] & used[1:])[0] + start
+        stops = np.where(used[:-1] & ~used[1:])[0] + start
+
+        data = np.zeros((len(sel), np.sum([b - a for a, b in zip(starts,
+                                                                 stops)])))
+        times = np.zeros(data.shape[1])
+        idx = 0
+        for start, stop in zip(starts, stops):  # get the data
+            if start == stop:
+                continue
+            end = idx + stop - start
+            data[:, idx:end], times[idx:end] = self[sel, start:stop]
+            idx = end
+
+        return data, times
```

Comment:
NaN option would be nice - as it would propagate across tfr for example, right? I used NaNs for this before because I had long epochs (~30 s) and rejecting whole epoch just because 1 s was bad seemed to be a waste of data.